### PR TITLE
Pay/Receive redesign

### PR DESCRIPTION
**Pull request by tomasz-ludek**
- DMD-1100 Send Payment | Enter Amount
- DMD-1101 Receive Dash | Show QR Code
- DMD-1124 Send Payment | Confirm Transaction Details

### CODE REVIEW

File: wallet/res/layout/send_coins_fragment1.xml [hunk 1]
```diff
@@ -0,0 +1,246 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (HashEngineering) [inline wallet/res/layout/send_coins_fragment1.xml]: Is this file used?  It has a 1 in the name.

File: wallet/src/de/schildbach/wallet/service/BlockchainServiceImpl.java [hunk 2]
```diff
@@ -111,6 +111,7 @@
 import android.os.PowerManager;
 import android.os.PowerManager.WakeLock;
 import androidx.core.app.NotificationCompat;
+import androidx.lifecycle.LifecycleService;
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/service/BlockchainServiceImpl.java]: This is interesting.  Perhaps later we can upgrade BlockchainService based on Bitcoin Wallet 6.x+

File: wallet/src/de/schildbach/wallet/ui/preference/PreferenceActivity.java [hunk 3]
```diff
@@ -68,7 +68,7 @@ private void initToolbar()
     @Override
     public void onBuildHeaders(final List<Header> target) {
         loadHeadersFromResource(R.xml.preference_headers, target);
-        hidePersonalSettingsIfLocked(target);
+//        hidePersonalSettingsIfLocked(target);
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/preference/PreferenceActivity.java]: This is acceptable for now until the WalletLock functions are integrated.

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java [hunk 4]
```diff
@@ -821,204 +379,43 @@ public void run() {
...
+        } else if (viewModel.paymentIntent.mayEditAmount()) {
+            return enterAmountSharedViewModel.hasAmount();
+        } else {
+            return viewModel.paymentIntent.hasAmount();
+        }
     }
 
     private boolean everythingPlausible() {
-        return state == State.INPUT && isPayeePlausible() && isAmountPlausible() && isPasswordPlausible();
+        return viewModel.state == SendCoinsViewModel.State.INPUT && isPayeePlausible() && isAmountPlausible();// && isPasswordPlausible();
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: comment left in

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java [hunk 5]
```diff
@@ -1221,501 +586,215 @@ protected void onFailure(Exception exception) {
...
-            case REGULAR_PAYMENT: {
-                sendRequest.coinSelector = ZeroConfCoinSelector.get();
-                sendRequest.useInstantSend = false;
-                sendRequest.feePerKb = fees.get(feeCategory);
-                break;
-            }
-        }
-
-        sendRequest.ensureMinRequiredFee = forceEnsureMinRequiredFee ? true : sendRequest.useInstantSend;
+        sendRequest.coinSelector = llmqInstantSendActive ? ZeroConfCoinSelector.get() : InstantXCoinSelector.get();
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: if llmqInstantSendActive is false, the `ZeroConfCoinSelector` can still be used.

We shouldn't use `InstantXCoinSelector` anymore

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java [hunk 6]
```diff
@@ -341,33 +156,34 @@ public void run() {
...
-                            setState(State.SENT);
+                            setState(SendCoinsViewModel.State.SENT);
 
                             // Auto-close the dialog after a short delay
                             if (config.getSendCoinsAutoclose()) {
                                 handler.postDelayed(new Runnable() {
                                     @Override
                                     public void run() {
-                                        activity.finish();
+//                                        activity.finish();
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: comment left in

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java (hunk 5 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: There are no rules that help qualify a transaction for instantsend anymore, except that if the transactions are chained then there has to be less than 24 in the chain.  In this case the chain is a list set of transactions that are spending each others outputs, but none of them are confirmed in a block yet.  I am not sure if we can determine this easily.

File: wallet/res/layout/send_coins_fragment1.xml (hunk 1 above)
Comment (tomasz-ludek) [reply wallet/res/layout/send_coins_fragment1.xml]: File removed

File: wallet/src/de/schildbach/wallet/service/BlockchainServiceImpl.java (hunk 2 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/service/BlockchainServiceImpl.java]: Would be awesome. They replaced some broadcasts with more convenient LiveDatas.

File: wallet/src/de/schildbach/wallet/ui/preference/PreferenceActivity.java (hunk 3 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/preference/PreferenceActivity.java]: I left this line commented out in order to simplify testing process.

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java (hunk 4 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: Removed

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java (hunk 5 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: Thanks for the explanation!

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java (hunk 6 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: Comment removed

**Comment by HashEngineering** [comment]
The travis tests fail with this:
```
Task :wallet:compileProdReleaseKotlin FAILED
e: /home/travis/build/HashEngineering/dash-wallet/wallet/src/de/schildbach/wallet/ui/SetPinActivity.kt: (119, 56): Object is not abstract and does not implement abstract member public abstract fun onBack(longClick: Boolean): Unit defined in de.schildbach.wallet.ui.widget.NumericKeyboardView.OnKeyboardActionListener
e: /home/travis/build/HashEngineering/dash-wallet/wallet/src/de/schildbach/wallet/ui/SetPinActivity.kt: (138, 13): 'onBack' overrides nothing
```

**Comment by HashEngineering** [comment]
The items mentioned in my last review were addressed.

I started the travis test again to see if it will pass.  The app will build and run without crashing.

**Comment by tomasz-ludek** [comment]
> A few comments were left on:
> 
> 1. Comments
> 2. Coin Selector
> 
> But there are also bugs with:
> 3. rounding fiat currency values
> 4. showing invalid payment request dialog on the Payment screen (might be krip phone/android 8.1 only -- if it is, then we can create a separate story for this issue to fix it)
> 5. Pressing MAX places the dash balance in the top field, even if the TOP field is local currency (MAX -> switch -> MAX).

Fixed

File: wallet/src/de/schildbach/wallet/ui/widget/ReceiveInfoView.kt [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        qrCodeBitmap = BitmapDrawable(resources, Qr.bitmap(paymentRequestUri))
+        qrCodeBitmap.isFilterBitmap = false
+
+        qr_preview.setImageDrawable(qrCodeBitmap)
+        address_preview.text = address.toBase58()
+    }
+
+    private fun refreshData() {
+        val walletApplication = context.applicationContext as WalletApplication
+        address = walletApplication.wallet.currentReceiveAddress()
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/widget/ReceiveInfoView.kt]: This should use wallet.freshReceiveAddress() to prevent address reuse.

Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/widget/ReceiveInfoView.kt]: Good find!
Fixed.

**Pull request merged**